docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,164 @@
+| Simulator or Soak Test PR(s)   | _TBD_                                            |
+
+### Background
+#### Description
+
+- Currently, the job upload can only be performed by hz-cli command. This command requires a JVM on the host machine. The purpose of this feature is to enable non-java clients to be able to upload jobs to Jet server 
+
+
+#### Terminology
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| Client Binary Protocol | The binary messages used between Hazelcast Clients and Hazelcast Server |
+
+#### Actors and Scenarios
+
+Any language that implements Client Binary Protocol
+
+### Functional Design
+#### Summary of Functionality
+
+Provide a list of functions user(s) can perform.
+- Upload a jar file and pass the same parameters that can be used with hz-cli
+
+#### Additional Functional Design Topics
+
+None
+
+##### Notes/Questions/Issues
+
+- The new functionality requires that all resources accessed by the uploaded job to be available on the server. For example a text file that is used to populate an IMap needs to be available on the server or within the uploaded jar. Because a server side job can not access client side resources.  
+- The client should send multi parts in a sequential manner. Out of order messages are not handled since it would require more resources on the server.
+  
+
+### User Interaction
+#### API design and/or Prototypes
+
+```java
+HazelcastInstance client = ...;
+JetService jetService = client.getJet();
+List<String> jobParameters = emptyList();
+
+String job1 = "job1";
+//If there is an error, throws JetException
+jetService.submitJobJar(getJarPath(),
+        null,
+        job1,
+        null,
+        jobParameters);
+```
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void submitJobJar(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
+```
+
+### Technical Design
+
+The client protocol needs to support job uploading. So that non-java clients can also upload jet jobs.
+For this purpose two new messages are added to client protocol
+
+1. uploadJobMetaData
+2. uploadJobMultipart
+
+**1. uploadJobMetaData Message**
+
+The upload process starts with uploadJobMetaData. This message contains
+
+| Term          | Type        | Definition                                                |
+|---------------|-------------|-----------------------------------------------------------|
+| sessionId     | UUID        | The UUID. This field associates all messages in a session |
+| filename      | String      | Name of the jar file without extension                    |
+| sha256Hex     | String      | Hexadecimal SHA256 of the jar file                        |
```

Comment:
I think it would be better to move `sha256Hex` to individual parts, so the client doesn't have to produce the hash (hence read all the file) before starting to upload parts.

Reply:
We want to check the integrity of the whole jar file. That's why we need this field in the meta data message. Individual part's  integrity is guaranteed by TCP, however the processing logic on the member may be faulty, disk maybe faulty etc. Moving this field to individual part messages does not give what we want. If you meant copying this field to individual part messages, I think TCP should give us sufficient protection

Reply:
It is possible that there isn't a direct TCP connection between the client and the server, there can be proxies, load balancers, etc. TCP guarantees wouldn't protect the integrity of the part in those cases.

Reply:
S3 upload API seems to have checksums for parts too: https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/upload-part.html
The relevant part of the doc:
```
upload-part
[--body <value>]
--bucket <value>
[--content-length <value>]
[--content-md5 <value>]
[--checksum-algorithm <value>]
[--checksum-crc32 <value>]
[--checksum-crc32-c <value>]
[--checksum-sha1 <value>]
[--checksum-sha256 <value>]
--key <value>
--part-number <value>
```

Reply:
Ok. I will add sha256 to parts too